### ISSUE TITLE
Disable automatically adding |variable| when I use the do keyword

### ISSUE THREAD

**Issue by zanloy** [atom/language-ruby#107]
This is just annoying. If I want to create a code block without a variable, I have to hit backspace to remove "|variable|" that it's trying to add and then hit enter which adds 2 lines inside the block. Then I have to spend time away from the keys to go and delete the extra line. This is a serious problem and should either be disabled all together or at a minimum given an option to disable this "feature".

**Comment by mnquintana** [comment]
Just to clarify, are you asking that the snippet be changed? The insertion of `do |variable| end` comes from a snippet that ships with this package - are you asking that it be changed to just `do end`? Or for it to be removed altogether? Because the latter is extremely unlikely to happen, because you can avoid expanding snippets by just avoiding pressing <kbd>Tab</kbd> or <kbd>Enter</kbd> or pressing <kbd>Esc</kbd> to cancel out of autocomplete.

**Comment by zanloy** [comment]
It's a worthless feature and promotes bad code. You should never use the generic |variable| to begin with. And if I hit tab, enter, or esc it still adds too many spaces that have to be manually adjusted.

**Comment by mnquintana** [comment]
I suggest you read up on [how snippets in Atom work](https://atom.io/docs/latest/using-atom-snippets) – when you expand a snippet (like the one associated with `do`, in this case), it also creates placeholders with tab stops, and preselects the first one with the intention that *you will change the placeholder text*. So indeed, using the name `variable` in a block in real code would be a terrible practice, but it is a placeholder name, not a suggestion for code style. 

As I mentioned above, if you don't want the snippet to be expanded, then you can press <kbd>Esc</kbd> to prevent that, but *only* if the snippet hasn't already been expanded.

I hope that helps!

**Comment by zanloy** [comment]
The problem with this and the proposed solution is that neither is realistic. When coding ruby, if I'm writing a code block I know has variables then I type "do |variable|" and then hit enter. When I do this it doesn't match and doesn't automatically create the "end" for the block. But if I'm writing a block that does not have variables then it automatically added "|variable|" to the end of my line which is annoying at best. The user knows if the block has variables or not so why does the module automatically assume that every code block will have a variable? I code exclusively in ruby and have never once needed the automatic addition. Every time now I have to move my fingers from the home row to backspace and hit the down arrow.

The problem with adding the dop snippet is that it's not intuitive to any ruby developer. Again if I know a code block has variables I've already typed |variable| and it doesn't match either snippet.

**Comment by zanloy** [comment]
The real solution would be to not use ruby keywords for snippets. As the text editor, you don't want to be intrusive or counter intuitive to the user. Adding functionality like these snippets is one thing but actually changing the behaviour seems more like a hindrance to the user.

A good example of this is writing rspec test cases. Here is an example from my rails app.

```ruby
require 'rails_helper'

RSpec.describe "users/show.html.slim", type: :view do
  before(:each) do
    @user = assign(:user, create(:user))
    @assigned_tickets = assign(:tickets, create_pair(:ticket, assignee: @user))
    render
  end

  it 'shows the user name' do
    expect(rendered).to match(@user.name)
  end

  it 'shows the user email' do
    expect(rendered).to match(@user.email)
  end
  
  it 'shows assigned tickets' do
    @assigned_tickets.each do |ticket|
      expect(rendered).to match(ticket.title)
    end
  end
end
```

Not a simple one of these code blocks takes variables but every time I hit `Enter` it tries to add a variable. This is just an example. Writing sinantra apps also have a lot of `get '/' do` type blocks that will never have variables.

**Comment by moklett** [comment]
@zanloy I agree that the extra newlines are annoying.  They _can_ be avoided however, once you re-train the brain to snippet navigation:

---

**`do ... end` block:**

1. Type `do` <kbd>tab</kbd> (` |variable|` will be selected automatically)
2. <kbd>del</kbd> (to erase the selection)
3. <kbd>tab</kbd> (to move the cursor inside the block - not <kbd>enter</kbd>!)

Results in:

```rb
do
  # Cursor will be here, without an extra newline
end
```

---

**`do |variable| ... end` block:**

1. Type `do` <kbd>tab</kbd> (` |variable|` will be selected automatically)
2. <kbd>tab</kbd> (to move the selection to the variable name, inside the pipes)
3. Type your variable name, e.g. `item` (`variable` is replaced)
4. <kbd>esc</kbd> (to indicate we're done typing the variable name)
5. <kbd>tab</kbd> (to move the cursor inside the block)

Results in:

```rb
do |item|
  # Cursor will be here, without an extra newline
end
```

---

FYI, I landed here trying to figure out how to avoid the extra blank lines.  I ended up figuring it out myself, so I thought I'd share.

**Comment by matt-may** [comment]
+1

**Comment by Jwan622** [comment]
How do I turn this off!

**Comment by perlun** [comment]
Totally agree, the feature as it is right now is annoying. Like someone said, **snippet names should not collide with language-level keywords**. The problem is that the `dop` snippet comes first, before the `do` template (without parameter), so if you just type Ruby code like you normally do (`foo do` &lt;enter&gt;) you get an incorrect template that you have to manually remove.

@kevinsawicki - why did you close the original issue? I think this should either be:

- changed to a more sensible default (i.e. don't add variables in my code that I haven't asked for :wink:) _or_
- made configurable, so its at least possible to change the default behavior.

I suggest we re-open the issue.

**Comment by 50Wliu** [comment]
@perlun The issue you are describing is different and has been fixed in Atom 1.6.0.

**Comment by perlun** [comment]
> @perlun The issue you are describing is different and has been fixed in Atom 1.6.0. There was an issue with alphabetical sorting.

Thanks for pointing that out - it actually works better with 1.6.0.

How can I turn off these snippets entirely? I tend to prefer writing the code instead of letting the editor do it for me... :stuck_out_tongue_winking_eye:

**Comment by 50Wliu** [comment]
You should be able to disable the `snippets` package.
